File: crates/runvault/src/status.rs
//! `status.json` — written last, with an atomic rename.
//!
//! A run directory without this file is a run that did not complete.

use serde::{Deserialize, Serialize};

/// How the run ended.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "schema-gen", derive(schemars::JsonSchema))]
#[serde(rename_all = "lowercase")]
pub enum State {
    /// `finish()` was called and the shallow checks passed.
    Finished,
    /// The process died, or `finish()` found the run inconsistent.
    Failed,
}

/// Why the run failed. Required when `state` is `failed`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[cfg_attr(feature = "schema-gen", derive(schemars::JsonSchema))]
#[serde(deny_unknown_fields)]
pub struct StatusError {
    /// A short, stable category (`dropped` / `verify` / an experiment's own kind).
    pub kind: String,
    /// The message a human reads.
    pub message: String,
}

/// How much was recorded. A cheap sanity check for a reader.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[cfg_attr(feature = "schema-gen", derive(schemars::JsonSchema))]
#[serde(deny_unknown_fields)]
pub struct Counts {
    /// Rows in `metrics.csv`.
    pub metrics: u64,
    /// Lines in `events.jsonl`.
    pub events: u64,
    /// Rows in `manifest.csv`.
    pub artifacts: u64,
}

/// `status.json`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[cfg_attr(feature = "schema-gen", derive(schemars::JsonSchema))]
#[serde(deny_unknown_fields)]
pub struct RunStatus {
    /// Version of `schema/v1`.
    pub schema_version: String,
    /// The run this status belongs to.
    pub run_uid: String,
    /// Finished or failed.
    pub state: State,
    /// When the run started.
    pub started_at: String,
    /// When it ended.
    pub finished_at: String,
    /// Wall-clock duration. The single source of truth; `metrics.csv` does not repeat it.
    pub duration_sec: f64,
    /// Process exit code. `0` or absent for a finished run.
    pub exit_code: Option<i64>,
    /// Which `-N` suffix the directory name needed, when `run_slug` collided.
    pub collision_index: Option<u64>,
    /// Why it failed.
    pub error: Option<StatusError>,
    /// How much was recorded.
    pub counts: Option<Counts>,
}

File: python/src/runvault/status.py
"""`status.json` — written last, with an atomic rename.

A run directory without this file is a run that did not complete.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

__all__ = ["Counts", "FAILED", "FINISHED", "RunStatus", "StatusError"]

#: `finish()` was called and the shallow checks passed.
FINISHED = "finished"
#: The process died, or `finish()` found the run inconsistent.
FAILED = "failed"


@dataclass(frozen=True)
class StatusError:
    """Why the run failed. Required when `state` is `failed`."""

    kind: str
    message: str

    def to_json(self) -> dict[str, Any]:
        return {"kind": self.kind, "message": self.message}


@dataclass
class Counts:
    """How much was recorded. A cheap sanity check for a reader."""

    metrics: int = 0
    events: int = 0
    artifacts: int = 0

    def to_json(self) -> dict[str, Any]:
        return {"metrics": self.metrics, "events": self.events, "artifacts": self.artifacts}


@dataclass(frozen=True)
class RunStatus:
    """`status.json`. Every field is written, so a reader never guesses a default."""

    schema_version: str
    run_uid: str
    state: str
    started_at: str
    finished_at: str
    duration_sec: float
    exit_code: int | None = None
    collision_index: int | None = None
    error: StatusError | None = None
    counts: Counts | None = None

    def to_json(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "run_uid": self.run_uid,
            "state": self.state,
            "started_at": self.started_at,
            "finished_at": self.finished_at,
            "duration_sec": self.duration_sec,
            "exit_code": self.exit_code,
            "collision_index": self.collision_index,
            "error": None if self.error is None else self.error.to_json(),
            "counts": None if self.counts is None else self.counts.to_json(),
        }
